### PR TITLE
Authorization

### PR DESCRIPTION
**Pull request by derekcollison**
This allows authorization of subjects for publish and subscribe filters with enforcement.

/cc @kozlovic @ColinSullivan1

### CODE REVIEW

File: server/errors.go [hunk 1]
```diff
@@ -1,19 +1,22 @@
...
 
-	// ErrAuthTimeout represents error condition on failed authorization due to timeout.
+	// ErrAuthTimeout represents an error condition on failed authorization due to timeout.
 	ErrAuthTimeout = errors.New("Authorization Timeout")
 
-	// ErrMaxPayload represents error condition when the payload is too big.
+	// ErrMaxPayload represents an error condition when the payload is too big.
 	ErrMaxPayload = errors.New("Maximum Payload Exceeded")
+
+	// ErrReservedPublish represents an error condition when the payload is too big.
```
Comment (kozlovic) [inline server/errors.go]: Typo, should be `ErrReserverdPublishSubject` and wrong content too.

File: server/client.go [hunk 2]
```diff
@@ -591,6 +640,17 @@ func (c *client) processSub(argo []byte) (err error) {
 		return nil
 	}
 
+	// Check permissions if applicable.
+	if c.perms != nil && c.perms.sub != nil {
```
Comment (kozlovic) [inline server/client.go]: As of now, it looks like if `c.perms` is not `nil`, then `perms.sub` and `perms.pub` will be not `nil`..

File: server/client.go [hunk 3]
```diff
@@ -146,6 +159,36 @@ func (c *client) initClient() {
...
+	if user.Permissions == nil {
+		return
+	}
+
+	// Process Permissions and map into client connection structures.
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.perms = &permissions{}
+	c.perms.sub = NewSublist()
```
Comment (kozlovic) [inline server/client.go]: No big deal, but should we create those only if there are permissions for sub and/or pub? I see that we later always check if `perms.sub` or `perms.pub` is not `nil` even when `perms` is not.

File: server/client.go [hunk 4]
```diff
@@ -814,9 +874,51 @@ func (c *client) processMsg(msg []byte) {
...
+		allowed, ok := c.perms.pcache[string(c.pa.subject)]
+		if ok && !allowed {
+			c.pubPermissionViolation(c.pa.subject)
+			return
+		}
+		if !ok {
+			r := c.perms.pub.Match(string(c.pa.subject))
+			if len(r.psubs) == 0 {
+				c.pubPermissionViolation(c.pa.subject)
+				c.perms.pcache[string(c.pa.subject)] = false
```
Comment (kozlovic) [inline server/client.go]: Are we not potentially growing the cache unbound here? If I understand, we reach this code when there is no cached permission, and we happen to not have permission for this subject. Suppose we arrive here with, say, `foo.1`, `foo.2`, etc.. and the user does not have permission for any matching, I think we will end-up here and keep adding to the cache?

**Comment by kozlovic** [comment]
Done with the review, I have a doubt about a possible unbounded growth of the pub cache. Otherwise, LGTM.

**Comment by derekcollison** [comment]
Thanks @kozlovic, agree on cache growth, already added those checks in. Will look more closely at the other comments as well.

File: server/opts.go [hunk 5]
```diff
@@ -71,13 +78,15 @@ type Options struct {
 	TLSConfig          *tls.Config   `json:"-"`
 }
 
+// Configuration file  quthorization section.
```
Comment (ColinSullivan1) [inline server/opts.go]: typo

File: server/opts.go [hunk 6]
```diff
@@ -325,17 +334,134 @@ func parseAuthorization(am map[string]interface{}) (*authorization, error) {
...
+	var subjects []string
+	switch v.(type) {
+	case string:
+		subjects = append(subjects, v.(string))
+	case []string:
+		subjects = v.([]string)
+	case []interface{}:
+		for _, i := range v.([]interface{}) {
+			subject, ok := i.(string)
+			if !ok {
```
Comment (mcqueary) [inline server/opts.go]: "cannot" is the preferred form in this context.

File: server/opts_test.go [hunk 7]
```diff
@@ -397,3 +397,90 @@ func TestMultipleUsersConfig(t *testing.T) {
 	}
 	processOptions(opts)
 }
+
+// Test highly depends on contents of the config file listed below. Any changes to that file
```
Comment (mcqueary) [inline server/opts_test.go]: well vs. weel.

File: server/opts.go [hunk 8]
```diff
@@ -325,17 +334,134 @@ func parseAuthorization(am map[string]interface{}) (*authorization, error) {
...
+				return nil, fmt.Errorf("Subject in permissions array can not be cast to string")
```
Comment (mcqueary) [inline server/opts.go]: "cannot" is the preferred form in this context.

File: server/opts_test.go [hunk 9]
```diff
@@ -397,3 +397,90 @@ func TestMultipleUsersConfig(t *testing.T) {
...
+// may very weel break this test.
```
Comment (mcqueary) [inline server/opts_test.go]: well vs. weel.

File: server/client.go (hunk 3 above)
Comment (ColinSullivan1) [inline server/client.go]: I agree, it'd simplify later checks as Ivan noted.

File: server/client.go [hunk 10]
```diff
@@ -932,6 +1045,11 @@ func (c *client) processMsg(msg []byte) {
 	}
 }
 
+func (c *client) pubPermissionViolation(subject []byte) {
+	c.sendErr(fmt.Sprintf("Permissions Violation for Publish to %q", subject))
+	c.Errorf("Publish Violation - User %q, Subject %q", c.opts.Username, subject)
```
Comment (ColinSullivan1) [inline server/client.go]: Could a badly behaved application flood server logs?

File: test/configs/authorization.conf [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+  # This is to benchmark pub performance.
+  BENCH = {
+    publish = "a"
+  }
+
+  # Just foo for testing
+  PASS: $2a$10$UHR6GhotWhpLsKtVP0/i6.Nh9.fuY73cWjLoJjb2sKT8KISBcUW5q
+
+  # Users listed with persmissions.
```
Comment (ColinSullivan1) [inline test/configs/authorization.conf]: permissions

File: test/client_auth_test.go [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+	// Test second user
+	url = fmt.Sprintf("nats://%s:%s@%s:%d/",
+		opts.Users[1].Username,
+		opts.Users[1].Password,
+		opts.Host, opts.Port)
+
+	nc, err = nats.Connect(url)
+	if err != nil {
+		t.Fatalf("Expected a succesful connect, got %v\n", err)
```
Comment (ColinSullivan1) [inline test/client_auth_test.go]: successful

**Comment by ColinSullivan1** [comment]
I'm wrapped up with the review as well.  Apart from the potential log growth issue, LGTM.

File: server/opts.go (hunk 5 above)
Comment (derekcollison) [inline server/opts.go]: fixed.

File: server/opts.go (hunk 8 above)
Comment (derekcollison) [inline server/opts.go]: fixed

File: server/opts_test.go (hunk 9 above)
Comment (derekcollison) [inline server/opts_test.go]: fixed, will add in automated spell check.. ;)

File: test/client_auth_test.go (hunk 12 above)
Comment (derekcollison) [inline test/client_auth_test.go]: fixed.

File: test/configs/authorization.conf (hunk 11 above)
Comment (derekcollison) [inline test/configs/authorization.conf]: fixed.

File: server/errors.go (hunk 1 above)
Comment (derekcollison) [inline server/errors.go]: fixed.

File: server/client.go (hunk 3 above)
Comment (derekcollison) [inline server/client.go]: Left as is but simplified checks in other code paths to just look for c.perms != nil.

File: server/client.go (hunk 2 above)
Comment (derekcollison) [inline server/client.go]: +!

File: server/client.go (hunk 4 above)
Comment (derekcollison) [inline server/client.go]: Code to prune is right below here on line 903.

File: server/client.go (hunk 10 above)
Comment (derekcollison) [inline server/client.go]: I was going back and forth between if this should be an error or a debug statement. When you try to connect to an authorized server its an error, so we kinda already have this issue in a different form already. Thoughts?

Comment (derekcollison) [inline server/client.go]: What I think we do is leave as an error, but I will mark a TODO to have thresholds where we cut off the client and/or blacklist them.

Comment (ColinSullivan1) [inline server/client.go]: I see your point about the connect, but in this case a simple user error (wrong subject) with typical API usage could end up being really harmful to the server.  How would you feel about making it a debug statement for now, then promoting it to an error when thresholds are implemented?

Comment (derekcollison) [inline server/client.go]: I had it as a debug statement.. Not sure. I need to start looking at the clients and make sure they easily do the right thing so we can avoid, my app doesn't work, etc..

**Comment by mcqueary** [comment]
LGTM

File: server/client.go (hunk 4 above)
Comment (kozlovic) [inline server/client.go]: Yes, but we return here. If a publisher publishes on foo.1, foo.2, foo.3, etc... or inboxes that are all unique you may have a case where the cache grows bigger than the set limit. Only after a positive match will you go to the cache pruning.

Comment (derekcollison) [inline server/client.go]: Good point.

Comment (derekcollison) [inline server/client.go]: fixed.

**Pull request merged**